http_call.go

```diff
@@ -24,6 +25,102 @@ import (
 	"sync"
 )
 
+type unaryHTTPCall struct {
+	httpClient       HTTPClient
+	onRequestSend    func(*http.Request)
+	validateResponse func(*http.Response) *Error
+
+	request  *http.Request
+	response *http.Response // nil until the request has been sent
+}
+
+func newUnaryHTTPCall(
+	ctx context.Context,
+	httpClient HTTPClient,
+	url *url.URL,
+	header http.Header,
+) *unaryHTTPCall {
+	// ensure we make a copy of the url before we pass along to the
+	// Request. This ensures if a transport out of our control wants
+	// to mutate the req.URL, we don't feel the effects of it.
+	url = cloneURL(url)
+
+	// This is mirroring what http.NewRequestContext did, but
+	// using an already parsed url.URL object, rather than a string
+	// and parsing it again. This is a bit funny with HTTP/1.1
+	// explicitly, but this is logic copied over from
+	// NewRequestContext and doesn't effect the actual version
+	// being transmitted.
+	request := (&http.Request{
+		Method:     http.MethodPost,
+		URL:        url,
+		Header:     header,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Body:       nil, // Set in Do.
+		Host:       url.Host,
+	}).WithContext(ctx)
+	return &unaryHTTPCall{
+		httpClient: httpClient,
+		request:    request,
+	}
+}
+
+// Do sends the request and waits for the response.
+func (u *unaryHTTPCall) Do(body *bytes.Buffer) *Error {
```

Comment:
Why do we need to split the operation into two phases -- `newUnaryHTTPCall` and then `call.Do`? Having mutators on the call, to support GET method, seems a little silly when the caller could just compute everything first and then construct the call with the correct method, body, and query params.

This split is needed for streaming (mainly client and bidi streams) so that the caller can define headers because actually sending the request. But that isn't the case for unary calls.

Reply:
I agree, was following the current pattern. I don't see why it is required for streaming though. Header map is provided on conn creation so  can be mutated without creating the request object. What am I missing?

Reply:
> Header map is provided on conn creation

This is not the case for client-stream and bidi endpoints. We create the conn and then return it in the form of a `ClientStreamForClient` or `BidiStreamForClient`. The user can then mutate headers via `stream.RequestHeader()` before sending via `stream.Send()`.

Reply:
It is the case for client-stream and bidi. 
- newDuplexHTTP creates the conn and passes in the header: https://github.com/connectrpc/connect-go/blob/b6d44ecfd51ddab5fa37fbbc7c3cbcece7940179/duplex_http_call.go#L59
- `stream.RequestHeader()` points to the map passed into the request object: https://github.com/connectrpc/connect-go/blob/b6d44ecfd51ddab5fa37fbbc7c3cbcece7940179/protocol_connect.go#L471

Reply:
On `stream.Send(nil)` we need to create the request object. For `RequestHeader()` we can return the `http.Header` that we would use to create the request object with and not mutate a request, like for unary as you suggested. I still don't see why we have to create the request object up front.

Reply:
Resolving as removed unary http call struct.